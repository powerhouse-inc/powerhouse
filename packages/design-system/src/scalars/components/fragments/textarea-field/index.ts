export { TextareaField, type TextareaFieldProps } from "./textarea-field.js";
