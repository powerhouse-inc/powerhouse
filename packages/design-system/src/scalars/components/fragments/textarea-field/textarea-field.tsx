import {
  Textarea,
  type TextareaProps,
} from "../../../../ui/components/data-entry/textarea/textarea.js";
import { type FieldErrorHandling, type WithDifference } from "../../types.js";
import { withFieldValidation } from "../with-field-validation/with-field-validation.js";

export type TextareaFieldProps = TextareaProps &
  FieldErrorHandling &
  WithDifference<string>;

const TextareaField = withFieldValidation<TextareaFieldProps>(Textarea);

TextareaField.displayName = "TextareaField";

export { TextareaField };
