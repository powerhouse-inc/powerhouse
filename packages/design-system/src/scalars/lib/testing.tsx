import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
