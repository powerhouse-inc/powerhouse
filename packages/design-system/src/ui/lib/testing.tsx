import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
