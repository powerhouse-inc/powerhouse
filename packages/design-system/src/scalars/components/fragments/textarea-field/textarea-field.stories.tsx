import { useState } from "react";
import type { Meta, StoryObj } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import { TextareaField } from "./textarea-field";

const meta = {
  title: "Document Engineering/Fragments/TextareaField",
  component: TextareaField,
  parameters: {
    layout: "centered",
    controls: {
      sort: "requiredFirst",
      expanded: true,
    },
  },
  tags: ["autodocs"],
  argTypes: {
    autoComplete: {
      control: "boolean",
      description:
        "Whether to enable browser autocompletion for the textarea field",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Behavior",
      },
    },
    autoExpand: {
      control: "boolean",
      description:
        "Whether the textarea should automatically expand with content",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Behavior",
      },
    },
    autoFocus: {
      control: "boolean",
      description:
        "Whether the textarea should automatically receive focus on mount",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Behavior",
      },
    },
    className: {
      control: "text",
      description: "Additional CSS classes to apply to the textarea",
      table: {
        type: { summary: "string" },
        category: "Styling",
      },
    },
    customValidator: {
      description:
        "Custom validation function that returns error message or null",
      table: {
        type: { summary: "(value: string) => string | null" },
        category: "Validation",
      },
    },
    defaultValue: {
      control: "text",
      description: "Initial value for uncontrolled component usage",
      table: {
        type: { summary: "string" },
        category: "State",
      },
    },
    description: {
      control: "text",
      description: "Helper text displayed below the textarea field",
      table: {
        type: { summary: "string" },
        category: "Content",
      },
    },
    disabled: {
      control: "boolean",
      description: "Whether the textarea field is disabled",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "State",
      },
    },
    errors: {
      control: "object",
      description:
        "Array of error messages to display below the textarea field",
      table: {
        type: { summary: "string[]" },
        defaultValue: { summary: "[]" },
        category: "Validation",
      },
    },
    id: {
      control: "text",
      description:
        "Unique identifier for the textarea field. Auto-generated if not provided",
      table: {
        type: { summary: "string" },
        category: "Technical",
      },
    },
    label: {
      control: "text",
      description:
        "Label text displayed above the textarea field for accessibility and UX",
      table: {
        type: { summary: "string" },
        category: "Content",
      },
    },
    lowercase: {
      control: "boolean",
      description:
        "Transforms input text to lowercase. Cannot be used with uppercase",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Text Transformation",
      },
    },
    maxLength: {
      control: "number",
      description: "Maximum number of characters allowed in the textarea",
      table: {
        type: { summary: "number" },
        category: "Validation",
      },
    },
    minLength: {
      control: "number",
      description: "Minimum number of characters required in the textarea",
      table: {
        type: { summary: "number" },
        category: "Validation",
      },
    },
    name: {
      control: "text",
      description:
        "Name attribute for the textarea field. Auto-generated if not provided",
      table: {
        type: { summary: "string" },
        category: "Technical",
      },
    },
    onChange: {
      action: "onChange",
      description: "Callback fired when the textarea value changes",
      table: {
        type: {
          summary: "(event: React.ChangeEvent<HTMLTextAreaElement>) => void",
        },
        category: "Events",
      },
    },
    pattern: {
      control: "text",
      description: "Regular expression pattern for input validation",
      table: {
        type: { summary: "string" },
        category: "Validation",
      },
    },
    placeholder: {
      control: "text",
      description: "Placeholder text displayed when the textarea is empty",
      table: {
        type: { summary: "string" },
        category: "Content",
      },
    },
    required: {
      control: "boolean",
      description: "Whether the textarea input is required",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Validation",
      },
    },
    rows: {
      control: "number",
      description: "Number of visible text lines in the textarea",
      table: {
        type: { summary: "number" },
        defaultValue: { summary: "3" },
        category: "Appearance",
      },
    },
    showErrorOnBlur: {
      control: "boolean",
      description:
        "Whether to display validation errors when the field loses focus",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Validation",
      },
    },
    showErrorOnChange: {
      control: "boolean",
      description: "Whether to display validation errors as the user types",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Validation",
      },
    },
    spellCheck: {
      control: "boolean",
      description: "Whether to enable browser spell checking on the textarea",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Behavior",
      },
    },
    trim: {
      control: "boolean",
      description:
        "Whether to remove leading and trailing whitespace from the value",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Text Transformation",
      },
    },
    uppercase: {
      control: "boolean",
      description:
        "Transforms input text to uppercase. Cannot be used with lowercase",
      table: {
        type: { summary: "boolean" },
        defaultValue: { summary: "false" },
        category: "Text Transformation",
      },
    },
    value: {
      control: "text",
      description:
        "Controlled value of the textarea field. Use with onChange handler",
      table: {
        type: { summary: "string" },
        category: "State",
      },
    },
    warnings: {
      control: "object",
      description: "Array of warning messages to display below the textarea",
      table: {
        type: { summary: "string[]" },
        defaultValue: { summary: "[]" },
        category: "Validation",
      },
    },
  },
} satisfies Meta<typeof TextareaField>;

export default meta;
type Story = StoryObj<typeof meta>;

// Basic Examples
export const Default: Story = {
  args: {
    placeholder: "Enter your text here...",
  },
};

export const WithLabel: Story = {
  args: {
    label: "Message",
    placeholder: "Enter your message",
  },
};

export const WithLabelAndDescription: Story = {
  args: {
    label: "Bio",
    description: "Tell us about yourself in a few sentences",
    placeholder: "I am...",
  },
};

// Validation States
export const Required: Story = {
  args: {
    label: "Comments",
    required: true,
    placeholder: "This field is required",
  },
};

export const WithSingleWarning: Story = {
  args: {
    label: "Feedback",
    value: "OK",
    warnings: ["Your feedback seems quite short"],
  },
};

export const WithMultipleWarnings: Story = {
  args: {
    label: "Feedback",
    value: "test",
    warnings: [
      "Consider providing more detailed feedback",
      "Your response may be too brief for proper evaluation",
    ],
  },
};

export const WithSingleError: Story = {
  args: {
    label: "Comments",
    value: "ab",
    errors: ["Comments must be at least 10 characters long"],
    minLength: 10,
  },
};

export const WithMultipleErrors: Story = {
  args: {
    label: "Comments",
    value: "ab",
    errors: [
      "Comments must be at least 10 characters long",
      "Comments must include a specific example",
    ],
    minLength: 10,
  },
};

export const WithWarningsAndErrors: Story = {
  args: {
    label: "Feedback",
    value: "test",
    warnings: [
      "Consider providing more detailed feedback",
      "Your response may be too brief",
    ],
    errors: [
      "Feedback must be at least 10 characters long",
      "Feedback must include specific examples",
    ],
  },
};

// Length Constraints
export const WithMinLength: Story = {
  args: {
    label: "Short answer",
    minLength: 20,
    placeholder: "Must be at least 20 characters",
    description: "Minimum length: 20 characters",
  },
};

export const WithMaxLength: Story = {
  args: {
    label: "Brief response",
    maxLength: 50,
    placeholder: "Maximum 50 characters allowed",
    description: "Maximum length: 50 characters",
  },
};

export const WithMinAndMaxLength: Story = {
  args: {
    label: "Constrained response",
    minLength: 20,
    maxLength: 100,
    placeholder: "Between 20 and 100 characters",
    description: "Please provide between 20 and 100 characters",
  },
};

// Visual States
export const Disabled: Story = {
  args: {
    label: "Disabled content",
    value: "This content cannot be edited",
    disabled: true,
  },
};

export const Focused: Story = {
  args: {
    label: "Focused textarea",
    placeholder: "This field is focused",
    autoFocus: true,
  },
  parameters: {
    pseudo: { focus: true },
  },
};

// Special Features
export const WithSpellCheck: Story = {
  args: {
    label: "Spell-checked textarea",
    spellCheck: true,
    placeholder: "Spell checking is enabled...",
    description: "This field will check your spelling",
  },
};

export const WithCustomRows: Story = {
  args: {
    label: "Custom height textarea",
    rows: 10,
    placeholder: "This textarea shows 10 rows...",
  },
};

export const WithCharacterCounter: Story = {
  args: {
    label: "Limited text",
    maxLength: 100,
    value: "This text shows the character counter in action",
    placeholder: "Start typing... (max 100 characters)",
  },
};

export const AutoExpanding: Story = {
  args: {
    label: "Auto-expanding textarea",
    autoExpand: true,
    placeholder: "This will grow as you type...",
    description:
      "The textarea will automatically expand as you type more content",
  },
};

// Text Transformations
export const WithTrimTransformation: Story = {
  args: {
    label: "Trimmed text",
    value: "  This text will have whitespace trimmed  ",
    trim: true,
    description: "Leading and trailing whitespace will be removed",
  },
};

export const WithUppercaseTransformation: Story = {
  args: {
    label: "Uppercase text",
    value: "This text will be uppercase",
    uppercase: true,
    description: "Text will be converted to uppercase",
  },
};

export const WithBothTransformations: Story = {
  args: {
    label: "Transformed text",
    value: "  this will be trimmed and uppercase  ",
    trim: true,
    uppercase: true,
    description: "Text will be trimmed and converted to uppercase",
  },
};

// Controlled
const ControlledTextarea = () => {
  const [value, setValue] = useState("");
  return (
    <TextareaField
      label="Controlled textarea"
      value={value}
      onChange={(e) => {
        setValue(e.target.value);
        action("onChange")(e);
      }}
      placeholder="Type something..."
      description="This is a controlled component example"
    />
  );
};

export const Controlled: Story = {
  render: () => <ControlledTextarea />,
  parameters: {
    docs: {
      source: {
        code: `
const ControlledTextarea = () => {
  const [value, setValue] = useState("");
  return (
    <TextareaField
      label="Controlled textarea"
      value={value}
      onChange={(e) => setValue(e.target.value)}
      placeholder="Type something..."
      description="This is a controlled component example"
    />
  );
};
`,
      },
    },
  },
};
