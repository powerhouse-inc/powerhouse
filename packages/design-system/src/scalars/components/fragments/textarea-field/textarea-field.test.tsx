import { screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../../lib/testing.js";
import { TextareaField } from "./textarea-field.js";

describe("TextareaField Component", () => {
  it("should match snapshot with props", () => {
    const { asFragment } = renderWithForm(
      <TextareaField
        name="textarea"
        label="Default Label"
        description="Default description"
        required
        maxLength={100}
        warnings={["Warning"]}
        errors={["Error"]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle basic props and attributes", () => {
    renderWithForm(
      <TextareaField
        name="textarea"
        label="Test Label"
        required
        disabled
        multiline
        rows={5}
        spellCheck={true}
      />,
    );

    const textarea = screen.getByRole("textbox");
    const label = screen.getByText("Test Label");

    expect(label.parentElement).toHaveTextContent("*");
    expect(textarea).toBeDisabled();
    expect(textarea).toHaveAttribute("rows", "5");
    expect(textarea).toHaveAttribute("spellcheck", "true");
  });

  it("should handle resize behavior", () => {
    const { rerender } = renderWithForm(<TextareaField name="textarea" />);
    let textarea = screen.getByRole("textbox");
    expect(textarea).toHaveClass("resize-y");

    rerender(<TextareaField name="textarea" autoExpand />);
    textarea = screen.getByRole("textbox");
    expect(textarea).toHaveClass("resize-none", "overflow-hidden");
  });

  describe("Messages and Feedback", () => {
    it("should handle warnings and errors", async () => {
      renderWithForm(
        <TextareaField
          name="textarea"
          description="Help text"
          warnings={["Warning 1", "Warning 2"]}
          errors={["Error 1", "Error 2"]}
        />,
      );

      expect(screen.getByText("Help text")).toBeInTheDocument();
      expect(screen.getByText("Warning 1")).toBeInTheDocument();
      expect(screen.getByText("Warning 2")).toBeInTheDocument();
      await waitFor(() =>
        expect(screen.getByText("Error 1")).toBeInTheDocument(),
      );
      await waitFor(() =>
        expect(screen.getByText("Error 2")).toBeInTheDocument(),
      );
    });
  });

  describe("Input Handling", () => {
    it("should handle various input methods and transformations", async () => {
      const user = userEvent.setup();

      renderWithForm(
        <TextareaField
          name="textarea"
          value="Initial"
          maxLength={20}
          uppercase
        />,
      );

      const textarea = screen.getByRole("textbox");

      // Regular typing
      await user.type(textarea, " text");
      expect(textarea).toHaveValue("INITIAL TEXT");

      // Emoji input
      await user.clear(textarea);
      await user.type(textarea, "👋🏻");
      expect(textarea).toHaveValue("👋🏻");

      // Paste handling
      await user.clear(textarea);
      // Create a new paste event using JSDOM's Event constructor
      const pasteEvent = new Event("paste", { bubbles: true });
      Object.defineProperty(pasteEvent, "clipboardData", {
        value: {
          getData: () => "Pasted",
        },
      });
      textarea.dispatchEvent(pasteEvent);

      // RTL text handling
      await user.clear(textarea);
      const rtlText = "مرحبا";
      await user.type(textarea, rtlText);
      expect(textarea).toHaveValue(rtlText);
    });
  });

  describe("Form Integration", () => {
    it("should handle form operations", async () => {
      const user = userEvent.setup();
      const handleSubmit = vi.fn((e: React.FormEvent) => e.preventDefault());

      renderWithForm(
        <form onSubmit={handleSubmit}>
          <TextareaField defaultValue="Initial" name="textarea" />
          <button type="submit">Submit</button>
          <button type="reset">Reset</button>
        </form>,
      );

      const textarea = screen.getByRole("textbox");

      // Submit
      await user.click(screen.getByRole("button", { name: "Submit" }));
      expect(handleSubmit).toHaveBeenCalled();

      // Reset
      await user.type(textarea, " modified");
      await user.click(screen.getByRole("button", { name: "Reset" }));
      expect(textarea).toHaveValue("Initial");
    });
  });
});
