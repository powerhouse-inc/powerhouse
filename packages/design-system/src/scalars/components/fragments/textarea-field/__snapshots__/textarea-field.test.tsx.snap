// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextareaField Component > should match snapshot with props 1`] = `
<DocumentFragment>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <label
        class="inline-flex items-center text-sm font-medium leading-4 text-gray-900 dark:text-gray-50 mb-[3px]"
        for=":r1:-textarea"
      >
        Default Label
        <span
          class="ml-1 text-gray-800 group-hover:text-gray-900 dark:text-gray-400 dark:group-hover:text-slate-50"
        >
          *
        </span>
      </label>
      <div
        class="relative"
      >
        <textarea
          aria-invalid="false"
          aria-required="true"
          class="flex min-h-9 w-full rounded-md text-sm font-normal leading-5 text-gray-900 dark:text-gray-50 dark:border-charcoal-700 dark:bg-charcoal-900 border border-gray-300 bg-white px-3 py-[7px] font-sans placeholder:text-gray-500 dark:placeholder:text-gray-600 focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0 focus-visible:ring-offset-white dark:focus-visible:ring-charcoal-300 dark:focus-visible:ring-offset-charcoal-900 dark:focus:bg-charcoal-900 focus:bg-gray-50 disabled:cursor-not-allowed disabled:border-gray-300 disabled:bg-gray-50 disabled:text-gray-700 disabled:dark:border-charcoal-800 disabled:dark:bg-charcoal-900 disabled:dark:text-gray-300 resize-y scrollbar-thin scrollbar-gutter-stable scrollbar-track-transparent scrollbar-thumb-gray-300 hover:scrollbar-thumb-gray-300 dark:scrollbar-thumb-charcoal-700 dark:hover:scrollbar-thumb-charcoal-700 scrollbar-thumb-rounded-md"
          id=":r1:-textarea"
          name="textarea"
          rows="3"
        />
        <div
          class="mt-0.5 flex justify-end -mb-1"
        >
          <div
            class="flex items-center text-[10px] leading-3"
          >
            <span
              class="text-gray-500"
            >
              0
            </span>
            <span
              class="text-gray-300 dark:text-gray-700"
            >
              /100
            </span>
          </div>
        </div>
      </div>
      <p
        class="font-sans text-sm font-normal leading-5 text-gray-600 dark:text-gray-500"
      >
        Default description
      </p>
      <p
        class="mb-0 inline-flex items-center text-xs font-medium text-yellow-900 dark:text-yellow-900"
        data-type="warning"
      >
        Warning
      </p>
    </div>
  </form>
</DocumentFragment>
`;
